extra_geom/detectors.py

```diff
@@ -854,6 +856,159 @@ def to_distortion_array(self, allow_negative_xy=False):
+        frontview : bool
+          If True (the default), x increases to the left, as if you were looking
+          along the beam. False gives a 'looking into the beam' view.
+        """
+        ax = super().inspect(axis_units=axis_units, frontview=frontview)
+        scale = self._get_plot_scale_factor(axis_units)
+
+        # Label modules and tiles
+        for ch, module in enumerate(self.modules):
+            quad, mod = (1 + v for v in divmod(ch, 4))
+            cx, cy, _ = module[4].centre() * scale
+            ax.text(cx, cy, f'Q{quad}M{mod}', fontweight='bold',
+                    verticalalignment='center',
+                    horizontalalignment='center')
+
+            for t in [0, 7]:
+                cx, cy, _ = module[t].centre() * scale
+                ax.text(cx, cy, 'T{}'.format(t + 1),
+                        verticalalignment='center',
+                        horizontalalignment='center')
+
+        ax.set_title(f'AGIPD-500K2G detector geometry ({self.filename})')
+        ax.set_aspect(1)
+        return ax
+
+    @staticmethod
+    def split_tiles(module_data):
+        # Split into 8 tiles along the slow-scan axis
+        # This simple slicing is faster than np.split().
+        return [module_data[..., s:s+64, :] for s in range(0, 512, 64)]
+
+    @classmethod
+    def _tile_slice(cls, tileno):
+        # Which part of the array is this tile?
+        # tileno = 0 to 7
+        tile_offset = tileno * cls.frag_ss_pixels
+        ss_slice = slice(tile_offset, tile_offset + cls.frag_ss_pixels)
+        fs_slice = slice(0, cls.frag_fs_pixels)  # Every tile covers the full 128 pixels
+        return ss_slice, fs_slice
+
+    @classmethod
+    def _module_coords_to_tile(cls, slow_scan, fast_scan):
+        tileno, tile_ss = np.divmod(slow_scan, cls.frag_ss_pixels)
+        return tileno.astype(np.int16), tile_ss, fast_scan
+
+    def to_distortion_array(self, allow_negative_xy=False):
+        """Return distortion matrix for AGIPD500K detector, suitable for pyFAI.
+
+        Parameters
+        ----------
+
+        allow_negative_xy: bool
+          If False (default), shift the origin so no x or y coordinates are
+          negative. If True, the origin is the detector centre.
+
+        Returns
+        -------
+        out: ndarray
+            Array of float 32 with shape (4096, 128, 4, 3).
+            The dimensions mean:
+
+            - 8192 = 8 modules * 512 pixels (slow scan axis)
+            - 128 pixels (fast scan axis)
+            - 4 corners of each pixel
+            - 3 numbers for z, y, x
+        """
+        # Overridden only for docstring
+        return super().to_distortion_array(allow_negative_xy)
+
+    def write_crystfel_geom(self, *args, **kwargs):
+        super().write_crystfel_geom(*args, nquads=1, **kwargs)
+
+    def position_modules_fast(self, *args, **kwargs):
+        data, _ = super().position_modules_fast(*args, **kwargs)
+        return data, (0, 0)  # center is the origin
```

Comment:
I don't think this is quite right. The aim of returning the centre here is to say where the origin of the detector layout you've given EXtra-geom falls *in the output array*. It doesn't make sense to hard-code it.

Reply:
I take now the center of the output array size

Reply:
I still think this is not quite right. I guess 'center' is a misnomer - the idea is to return the location corresponding to the origin in the geometry object, i.e. the PONI. If the caller wants `(height / 2, width / 2)`, they can easily get that from the array shape.

Reply:
Oh, yeah. I agree. I misunderstood what you meant, will change taht

Reply:
Hum, thinking about it

>  the idea is to return the location corresponding to the origin in the geometry object, i.e. the PONI.

In that case, it should always be `(0, 0)`, right. Or I did not get it... :thinking:

Reply:
Sorry, should have been more specific. It's the origin from the geometry object, in the coordinates of the output array. So, for this class, I'd expect the `from_origin(origin=` parameter - in pixel units from the low-x, low-y corner - to be returned as the centre position in the array as well (but reversed, because array indexing goes (y, x)).

The idea is to tell you where the theoretical origin/PONI is in the assembled image.

Reply:
Aha, thanks. Hope I got that right now :D